draft-ietf-ppm-dap.md

```diff
@@ -2117,29 +2110,43 @@ Helper knows it can clean up its state.
+one of the Aggregators fails during this process in a manner that results in it
+losing its state, then the Aggregators risk losing the data associated with the
+aggregation job, or in some cases all of the batches associated with the
+aggregation job (if one of the Aggregators considers the aggregation to be
+complete but the other does not, the Aggregators will encounter a
+`batchMismatch` error on attempting to collect the batches associated with the
+aggregation job). This section describes how Aggregators can recover from this
+scenario.
+
 `AggregationJobContinueReq` messages contain a `step` field, allowing
-Aggregators to ensure that their peer is on an expected step of DAP
-aggregation. In particular, the intent is to allow recovery from a scenario
-where the Helper successfully advances from step `n` to `n+1`, but its
-`AggregationJobResp` response to the Leader gets dropped due to something like
-a transient network failure. The Leader could then resend the request to have
-the Helper advance to step `n+1` and the Helper should be able to retransmit
-the `AggregationJobResp` that was previously dropped. To make that kind of
-recovery possible, Aggregator implementations SHOULD checkpoint the most recent
-step's prep state and messages to durable storage such that the Leader can
-re-construct continuation requests and the Helper can re-construct continuation
-responses as needed.
-
-When implementing an aggregation step skew recovery strategy, the Helper SHOULD
-ensure that the Leader's `AggregationJobContinueReq` message did not change when
-it was re-sent (i.e., the two messages must be identical). This prevents the
-Leader from re-winding an aggregation job and re-running an aggregation step
-with different parameters.
-
-One way the Helper could address this would be to store a digest of the Leader's
-request, indexed by aggregation job ID and step, and refuse to service a request
-for a given aggregation step unless it matches the previously seen request (if
-any).
+Aggregators to ensure that their peer is on the expected step of DAP
+aggregation. An aggregation job is said to be "skewed" if the Leader and Helper
+have a different view of the last completed step of aggregation; for example, if
+the Leader does not receive or successfully process an `AggregationJobResp` from
+the Helper, the Leader will believe that the last completed step is one less
+than the Helper. The intent of the `step` field is to allow detection and
+recovery from a skewed aggregation job. When the Leader receives a
+`stepMismatch` error, it SHOULD re-send the original `AggregationJobInitReq`
+message to restart the aggregation process and allow aggregation to complete
+successfully. To make this kind of recovery possible, Leader implementations
+SHOULD retain enough information to reconstruct the `AggregationJobInitReq`
+message in the case that the aggregation process becomes desynchronized.
```

Comment:
I think this SHOULD is redundant, and maybe can be cut completely? What's the intent of this sentence?
```suggestion
successfully.
```

Reply:
I'm explicitly suggesting/hinting to Leader implementations that they need to keep the original request around (or enough information to reconstruct it byte-for-byte), if they want to implement skew recovery.

I suppose this is implied well enough by the previous sentence, and the following description of the Helper behavior -- I removed this sentence.